account_invoice_discount_lines/tests/test_invoice_discount_lines.py

```diff
@@ -0,0 +1,132 @@
+# © 2017 Opener BV (<https://opener.amsterdam>)
+# © 2020 Vanmoof BV (<https://www.vanmoof.com>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+from odoo import tests
+from odoo.tests.common import TransactionCase
+
+
+@tests.tagged('standard', 'at_install')
+class TestInvoiceDiscountLines(TransactionCase):
+    def setUp(self):
+        super(TestInvoiceDiscountLines, self).setUp()
+        self.tax = self.env['account.tax'].create({
+            'name': 'Test tax',
+            'amount': .1,
+            'amount_type': 'percent',
+            'account_id': self.env.ref('l10n_generic_coa.1_conf_iva').id,
+            'refund_account_id': self.env.ref('l10n_generic_coa.1_conf_iva').id,
+        })
+        product = self.env.ref('product.product_product_2')
+        self.env.user.company_id.write({'discount_product': product.id})
+        account = self.env.ref('l10n_generic_coa.1_conf_a_recv')
+        product.write({
+            'name': 'Discount',
+            'property_account_income': account.id,
+            'property_account_expense': account.id,
+        })
+        self.journal = self.env['account.journal'].search(
+            [('name', '=', 'Bank')], limit=1)
+        self.journal.currency_id = self.env.ref(
+            'base.main_company').currency_id.id
+
+    def create_invoice(self, price_unit=100, discount=10):
+        invoice = self.env['account.invoice'].create({
+            'account_id': self.env.ref('l10n_generic_coa.1_conf_a_recv').id,
+            'journal_id': self.journal.id,
+            'partner_id': self.env.ref('base.res_partner_12').id,
+            'name': 'Test invoice',
+            'invoice_line_ids': [
+                (0, 0, {
+                    'account_id': self.env.ref(
+                        'l10n_generic_coa.1_conf_a_sale').id,
+                    'name': '[PCSC234] PC Assemble SC234',
+                    'price_unit': price_unit,
+                    'quantity': 1,
+                    'discount': discount,
+                    'invoice_line_tax_id': [(4, self.tax.id)],
+                    'product_id': self.env.ref('product.product_product_3').id,
+                    'uos_id': self.env.ref('uom.product_uom_unit').id,
+                })],
+        })
+        invoice.compute_taxes()
+        return invoice
+
+    def test_01_invoice_discount_lines(self):
+        self.tax.account_collected_id = False
+        invoice = self.create_invoice()
+        self.assertEqual(len(invoice.invoice_line_ids), 1)
+        base_line = invoice.invoice_line_ids
+        self.assertTrue(
+            self.env.user.company_id.currency_id.is_zero(
```

Comment:
In the version 8 tests the amount was 99 https://github.com/VanMoof/account-invoicing/pull/1/commits/eeadbaaa9897f082b77c0e22e5a07abeeae3d43f#diff-4a95f0cfd40b3af02d5c39b446330838R74
As far as I understand this it should be 90. As 100 minus 10% discount is 90. Maybe I am missing.
This assertion has been repeated  in all of the tests below.

Reply:
I think the 90 euros is the net amount. The additional 9 euros is the 10% tax.

Reply:
OK than I don't have the tax in the demo setup. Should I change the test so we get tax on the line?

Reply:
Yes please! I added the tax to the test on purpose, to make sure that the discount lines mechanism does not mess up the tax amount.

Reply:
Tax was there I was writing it to a wrong field  "invoice_line_tax_id" vs "invoice_line_tax_ids" now I get different numbers. Here for tax .1 total value is 90.09 - seams correct to me but not 99.0 like the original test.
![image](https://user-images.githubusercontent.com/45589656/82919530-75d3e600-9f76-11ea-96e6-0ab4092cc887.png)

After applying the discount line we have total value 90.10 because tax was applied on the 100 USD line.
![image](https://user-images.githubusercontent.com/45589656/82919834-d2cf9c00-9f76-11ea-8881-9b6c45a3d86d.png)

Later on https://github.com/VanMoof/account-invoicing/pull/1/files#diff-4a95f0cfd40b3af02d5c39b446330838R106  the test fails because of a rounding is_zero() this is the difference 0.009999999999990905 the currency.rounding is 0.01. This is just wrong  0.009999999999990905  is less than 0.01.I guess 0.009999999999990905 is rounded up to 0.01 and the is_zero() fails.

Ill check the version 8.0 now to figure out how and why the old test passes.

Reply:
Update:
Test in the version 8 shows as follows
![image](https://user-images.githubusercontent.com/45589656/82925251-c00c9580-9f7d-11ea-8c06-ed0a64d44e4f.png)
That is tax of 10%
What I think happens is meaning of the tax amout field taht we set here https://github.com/VanMoof/account-invoicing/pull/1/files#diff-4a95f0cfd40b3af02d5c39b446330838R14 is changed. In 8 ".1" means 10% in 12 ".1" mens exactly what is says 0.1 %

Reply:
Wow. Odoo is really trying to piss us off aren't they, with changed name fields and changed semantics of amount precisions. But great that you tracked these changes! As for the taxes I have made a suggestion above.